tests/_utils/test_closure.py

```diff
@@ -228,3 +239,361 @@ def _recommend_book_series_from_available(genre: str) -> openai.OpenAIDynamicCon
+
+    def test_func():
+        pass
+
+    module = inspect.getmodule(test_func)
+    with patch.object(module, "__file__", None):
+        location = collector._get_code_location(test_func)
+        assert location.module_path == ""
+
+
+def test_is_user_defined_import():
+    """Test the _is_user_defined_import method."""
+    collector = _DependencyCollector()
+    assert collector._is_user_defined_import("from typing import List") is True
+    assert collector._is_user_defined_import("import openai") is False
+    assert collector._is_user_defined_import("import custom_module") is True
+
+
+def test_collect_decorator_calls_with_attribute():
+    """Test _collect_decorator_calls with attribute nodes."""
+
+    def decorator(func):
+        return func
+
+    class Decorators:
+        @staticmethod
+        def deco():
+            return decorator
+
+    @Decorators.deco()
+    def func():
+        pass
+
+    collector = _DependencyCollector()
+    collector.collect(func)
+    # No exception should be raised
+
+
+def test_visit_call_with_attribute():
+    """Test visit_Call handles attribute function calls."""
+
+    class SomeClass:
+        def method(self):
+            pass
+
+    def func():
+        obj = SomeClass()
+        obj.method()
+
+    closure, _ = compute_closure(func)
+    assert "SomeClass" in closure
+    assert "def func()" in closure
+
+
+def test_get_code_location_no_module():
+    """Test that _get_code_location handles functions with no module."""
+
+    def test_func():
+        pass
+
+    with patch("inspect.getmodule", return_value=None):
+        collector = _DependencyCollector()
+        with pytest.raises(ValueError):
+            collector._get_code_location(test_func)
+
+
+def test_collect_class_definition_no_source():
+    """Test handling class definition with no source."""
+    collector = _DependencyCollector()
+    module = Mock()
+
+    class TestClass:
+        pass
+
+    module.TestClass = TestClass
+    with (
+        patch("inspect.isclass", return_value=True),
+        patch("inspect.getsource", side_effect=TypeError),
+    ):
+        collector._collect_class_definition("TestClass", module)
+        # Should not raise exception
+
+
+def test_is_serializable_value():
+    """Test all cases of _is_serializable_value."""
+    collector = _DependencyCollector()
+    assert collector._is_serializable_value(42)
+    assert collector._is_serializable_value(3.14)
+    assert collector._is_serializable_value("string")
+    assert collector._is_serializable_value(True)
+    assert collector._is_serializable_value(None)
+    assert not collector._is_serializable_value(lambda x: x)
+    assert not collector._is_serializable_value(object())
+
+
+def test_collect_with_type_aliases():
+    """Test handling of type aliases in annotations."""
+    from typing import TypeAlias
+
+    MyType: TypeAlias = str  # pyright: ignore [reportGeneralTypeIssues]
+
+    def func(param: MyType) -> MyType:
```

Comment:
hmmmm, this is interesting

if the user were to switch the MyType alias from `str` to `int` then the closure and hash would remain the same even though they should change

the same is also true if the user updates from `MyType` to `str` which should have the same closure and hash but wont

I'm thinking for this case we should actually resolve the alias and use the original definition (i.e. `str` here) for the closure rather than the alias so that the closure/hash versioning is more consistent and correct

WDYT?